src/dlsnode/storage/BucketFile.d

```diff
@@ -440,6 +440,66 @@ public class BucketFile: OutputStream
 
     /**************************************************************************
 
+        Reads data from the file not blocking the fiber if there's not enough
+        data, giving that opportunity to the user. In that case, user should
+        suspend the fiber, wait for it to be resumed and call this method again.
+
+        Params:
+            waiting_context = ContextAwaitingJob used to wake
+                the fiber when the read is completed
+            buf = a buffer to write the data read from the file
+            wait_for_data = indicator if the fiber should suspend and wait until
+                            resumed via `waiting_context`
+
+        Returns:
+            number of bytes read from the file, or -1 if this should be suspended,
+            and retried again
+
+    **************************************************************************/
+
+    public ssize_t readData (ContextAwaitingJob waiting_context,
+            void[] buf, out bool wait_for_data)
+    {
+        assert(this.async_io);
+
+        // Let's first confirm that the last read didn't fail
+        if (this.aio_errno != 0)
```

Comment:
Hm, you're right, I need to think better about encapsulation at this point. Right now, the object is owned by the iterator, so the context is always the same. Perhaps it can pass back the "promise" object to the user. I'll see to make this better, you're right, this feels clunky.

Reply:
Yeah, something like that sounds like it'd be better.

Reply:
`BucketFile` isn't shared between requests?

Reply:
Only the one that does the writing, not the one that reads. We always open file and close file in the iterators.

Reply:
Ah ok. But it's shared between concurrent Put requests?

Reply:
Yes, that's right (that cache of open bucket files), but they don't read data.

Sharing for reading would be more problematic, as the underlying file descriptor would have to be read without changing the offsets, so only `pread` could be used. So, this also applies for the operating system file API - you can share the file for appending without thinking, but anything else is a dangerous zone.

Reply:
Ahh, so the same class is used in different ways for reading vs writing. I wonder if this could be asserted somehow.

Reply:
It probably could/should be.

Reply:
Would be nice. Currently it's purely down to convention (and possibly documentation -- I've not looked).

Reply:
Yeah, I'm not sure there's any documentation about this as well. Will add the enforcement, and then the documentation around it.